File: Iteration1/oosd_project/src/main/java/com/cse3063f25grp1/intent/IntentRulesLoader.java
package com.cse3063f25grp1.intent;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import com.cse3063f25grp1.model.Intent;

public class IntentRulesLoader {

    public Map<Intent, List<String>> loadRules(Path rulesFilePath) {
        try {
            List<String> lines = Files.readAllLines(rulesFilePath);
            return parseIntentRules(lines);
        } catch (IOException e) {
            throw new RuntimeException("Failed to load intent rules from: " + rulesFilePath, e);
        }
    }

    private Map<Intent, List<String>> parseIntentRules(List<String> lines) {
        Map<Intent, List<String>> rules = new LinkedHashMap<>();
        List<Intent> priority = new ArrayList<>();
        boolean inIntentPriority = false;
        boolean inKeywordRules = false;
        Intent currentIntent = null;
        List<String> currentKeywords = null;

        for (String line : lines) {
            String trimmed = line.trim();

            if (trimmed.isEmpty() || trimmed.startsWith("#")) {
                continue;
            }

            if (trimmed.equals("intent_priority:")) {
                inIntentPriority = true;
                inKeywordRules = false;
                continue;
            }

            if (trimmed.equals("keyword_rules:")) {
                inKeywordRules = true;
                inIntentPriority = false;
                continue;
            }

            if (inIntentPriority && trimmed.startsWith("-")) {
                String intentName = trimmed.substring(1).trim();
                if (intentName.startsWith("\"") && intentName.endsWith("\"")) {
                    intentName = intentName.substring(1, intentName.length() - 1);
                }
                Intent intent = parseIntentName(intentName);
                priority.add(intent);
            }

            if (inKeywordRules) {
                if (trimmed.endsWith(":") && !trimmed.startsWith("-")) {
                    if (currentIntent != null && currentKeywords != null) {
                        rules.put(currentIntent, currentKeywords);
                    }

                    String intentName = trimmed.substring(0, trimmed.length() - 1).trim();
                    currentIntent = parseIntentName(intentName);
                    currentKeywords = new ArrayList<>();
                } else if (trimmed.startsWith("-")) {
                    if (currentIntent != null && currentKeywords != null) {
                        String keyword = trimmed.substring(1).trim();
                        if (keyword.startsWith("\"") && keyword.endsWith("\"")) {
                            keyword = keyword.substring(1, keyword.length() - 1);
                        }
                        currentKeywords.add(keyword);
                    }
                }
            }
        }

        if (currentIntent != null && currentKeywords != null) {
            rules.put(currentIntent, currentKeywords);
        }

        // Priority'ye göre sırala
        Map<Intent, List<String>> orderedRules = new LinkedHashMap<>();
        for (Intent intent : priority) {
            if (rules.containsKey(intent)) {
                orderedRules.put(intent, rules.get(intent));
            }
        }

        // Priority'de olmayan intent'leri de ekle
        for (Map.Entry<Intent, List<String>> entry : rules.entrySet()) {
            if (!orderedRules.containsKey(entry.getKey())) {
                orderedRules.put(entry.getKey(), entry.getValue());
            }
        }

        return orderedRules;
    }

    private Intent parseIntentName(String yamlName) {
        switch (yamlName.toUpperCase()) {
            case "REGISTRATION":
                return Intent.Registration;
            case "STAFF_LOOKUP":
                return Intent.StaffLookup;
            case "POLICY_FAQ":
                return Intent.PolicyFAQ;
            case "COURSE":
                return Intent.Course;
            case "UNKNOWN":
                return Intent.Unknown;
            default:
                throw new IllegalArgumentException("Unknown intent name: " + yamlName);
        }
    }
}

File: iteration1_py/src/intent/intent_rules_loader.py
from typing import List, Dict, Optional
from pathlib import Path

from models import Intent


class IntentRulesLoader:

    def load_rules(self, rules_file_path: Path) -> Dict[Intent, List[str]]:
        try:
            lines: List[str] = rules_file_path.read_text(encoding="utf-8").splitlines()
            return self.__parse_intent_rules(lines)
        except Exception as e:
            raise RuntimeError(
                f"Failed to load intent rules from: {rules_file_path}"
            ) from e

    def __parse_intent_rules(self, lines: List[str]) -> Dict[Intent, List[str]]:

        rules: Dict[Intent, List[str]] = {}
        priority: List[Intent] = []

        in_intent_priority: bool = False
        in_keyword_rules: bool = False

        current_intent: Optional[Intent] = None
        current_keywords: Optional[List[str]] = None

        for line in lines:
            trimmed: str = line.strip()

            # Skip empty or commented lines
            if not trimmed or trimmed.startswith("#"):
                continue

            # Section switches
            if trimmed == "intent_priority:":
                in_intent_priority = True
                in_keyword_rules = False
                continue

            if trimmed == "keyword_rules:":
                in_keyword_rules = True
                in_intent_priority = False
                continue

            # Priority section
            if in_intent_priority and trimmed.startswith("-"):
                intent_name: str = trimmed[1:].strip()
                intent_name = self.__strip_quotes(intent_name)
                intent: Intent = self.__parse_intent_name(intent_name)
                priority.append(intent)
                continue

            # Keyword rules section
            if in_keyword_rules:
                if trimmed.endswith(":") and not trimmed.startswith("-"):
                    # Save previous entry
                    if current_intent is not None and current_keywords is not None:
                        rules[current_intent] = current_keywords

                    # New intent block
                    intent_name: str = trimmed[:-1].strip()
                    current_intent = self.__parse_intent_name(intent_name)
                    current_keywords = []
                    continue

                # Keyword line
                if trimmed.startswith("-") and current_intent is not None:
                    kw: str = trimmed[1:].strip()
                    kw = self.__strip_quotes(kw)
                    current_keywords.append(kw)

        # Save last block
        if current_intent is not None and current_keywords is not None:
            rules[current_intent] = current_keywords

        # Re-order rules based on priority
        ordered_rules: Dict[Intent, List[str]] = {}

        for intent in priority:
            if intent in rules:
                ordered_rules[intent] = rules[intent]

        # Add intents not in priority
        for intent, kw_list in rules.items():
            if intent not in ordered_rules:
                ordered_rules[intent] = kw_list

        return ordered_rules

    def __strip_quotes(self, text: str) -> str:

        if (text.startswith('"') and text.endswith('"')) or \
           (text.startswith("'") and text.endswith("'")):
            return text[1:-1]
        return text

    def __parse_intent_name(self, yaml_name: str) -> Intent:
    
        name = yaml_name.upper()

        if name == "REGISTRATION":
            return Intent.Registration
        if name == "STAFF_LOOKUP":
            return Intent.StaffLookup
        if name == "POLICY_FAQ":
            return Intent.PolicyFAQ
        if name == "COURSE":
            return Intent.Course
        if name == "UNKNOWN":
            return Intent.Unknown

        raise ValueError(f"Unknown intent name: {yaml_name}")
